service/notification/internal/app/scheduler/scheduler.go

```diff
@@ -150,20 +151,29 @@ func (s *dispatchScheduler) AddSubscription(sub *entities.Subscription) {
 	s.scheduledDispatches[sub.DispatchID] = dispatch
 }
 
-// func (s *dispatchScheduler) CancelSubscription(sub entities.Subscription) {
-// 	s.mu.Lock()
-// 	defer s.mu.Unlock()
-// 	schdledDispatch, ok := s.scheduledDispatches[sub.DispatchID]
-// 	if !ok {
-// 		return
-// 	}
-// 	emailIndex := slices.Index(schdledDispatch.Data.Emails, sub.Email)
-// 	if emailIndex < 0 {
-// 		return
-// 	}
-// 	countOfSubscribers := len(schdledDispatch.Data.Emails)
-// 	schdledDispatch.Data.Emails[emailIndex] = schdledDispatch.Data.Emails[countOfSubscribers-1]
-// 	schdledDispatch.Data.Emails = schdledDispatch.Data.Emails[:countOfSubscribers-1]
-// 	s.cron.Remove(schdledDispatch.entryID)
-// 	s.cron.AddJob(schdledDispatch.spec.String(), NewDispatchJob(schdledDispatch.Data, s.broker))
-// }
+func (s *dispatchScheduler) RemoveSubscriberFromDispatch(email, dispatchID string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	schdledDispatch, ok := s.scheduledDispatches[dispatchID]
+	if !ok {
+		return
+	}
+	emailIndex := slices.Index(schdledDispatch.Data.Emails, email)
+	if emailIndex < 0 {
+		return
+	}
+
+	s.cron.Remove(schdledDispatch.EntryID)
```

Comment:
What if you have 2 subs for the same dispatch and one of them unsubscribed. So here the second one will be unsubsribed too?

Reply:
Tthe job for the rest of subscribers is reschedulled at line 178 of this file

Reply:
Okay. Imagine you have millions of subs, so for every unsubscribe you will reschedule all of them?)
Seems like not great solution for this. Maybe you can rethink somehow your solution, so you won't need to reschedule others?

Reply:
I rewrited scheduler logic. Now it sets 1 permanent job, which is runned every 15 minutes. The job contains of next steps:
- fetches schedulled dispatches from store
- iterates fetched dispatches and send those that satisfy next condition:
   `dispatch.SendAt > now > dispatch.SendAt+15min`